File: Src/LoLMatchAccepterNet/LCU/LcuClient.cs
using LoLMatchAccepterNet.LCU;

namespace LeagueMatchAccepter
{
    public class LcuClient : IDisposable
    {
        private readonly LcuConnectionInfo? _connectionInfo;
        private readonly HttpClient? _client;
        private readonly MatchAcceptService? _matchAcceptService;
        private CancellationTokenSource? _cts;
        private bool disposedValue;

        public LcuClient()
        {
            var processDiscovery = new LeagueProcessDiscovery();
            _connectionInfo = processDiscovery.DiscoverLeagueClient();

            if (_connectionInfo?.IsValid == true)
            {
                var clientFactory = new LcuHttpClientFactory();
                _client = clientFactory.CreateClient(_connectionInfo);

                int port = int.Parse(_connectionInfo.Port);
                var game = new Game(_client, port);
                _matchAcceptService = new MatchAcceptService(game);
            }
        }

        public bool IsClientFound()
        {
            return _connectionInfo?.IsValid == true;
        }

        public async Task<bool> AutoAccept()
        {
            if (_matchAcceptService == null)
            {
                return false;
            }

            Console.WriteLine("Auto-accept running. Press ESC to exit.");
            Console.WriteLine("Waiting for match...");

            _cts = new CancellationTokenSource();
            var backgroundTask = Task.Run(() => RunAutoAcceptAsync(_cts.Token));

            while (!backgroundTask.IsCompleted)
            {
                if (Console.KeyAvailable && Console.ReadKey(true).Key == ConsoleKey.Escape)
                {
                    _cts.Cancel();
                    break;
                }
                await Task.Delay(1000);
            }

            try
            {
                return await backgroundTask;
            }
            catch (OperationCanceledException)
            {
                return true;
            }
        }

        private async Task<bool> RunAutoAcceptAsync(CancellationToken cancellationToken)
        {
            try
            {
                await _matchAcceptService!.RunAutoAcceptLoopAsync(cancellationToken);
                return true;
            }
            catch (OperationCanceledException)
            {
                throw;
            }
            catch (Exception)
            {
                return false;
            }
        }

        protected virtual void Dispose(bool disposing)
        {
            if (!disposedValue)
            {
                if (disposing)
                {
                    _cts?.Cancel();
                    _cts?.Dispose();
                    _client?.Dispose();
                }
                disposedValue = true;
            }
        }

        public void Dispose()
        {
            Dispose(disposing: true);
            GC.SuppressFinalize(this);
        }
    }
}

File: Src/LoLMatchAccepterCpp/src/LCU/LcuClient.cpp
#include "LcuClient.h"
#include <iostream>
#include <thread>
#include <conio.h>

LcuClient::LcuClient() : _cancelToken(false) {
    LeagueProcessDiscovery processDiscovery;
    _connectionInfo = processDiscovery.DiscoverLeagueClient();

    if (_connectionInfo.has_value() && _connectionInfo->IsValid()) {
        _client = std::make_shared<LcuHttpClient>(_connectionInfo.value());
        auto game = std::make_shared<Game>(_client);
        _matchAcceptService = std::make_shared<MatchAcceptService>(game);
    }
}

LcuClient::~LcuClient() {
    _cancelToken.store(true);
}

bool LcuClient::IsClientFound() const {
    return _connectionInfo.has_value() && _connectionInfo->IsValid();
}

bool LcuClient::AutoAccept() {
    if (!_matchAcceptService) {
        return false;
    }

    std::cout << "Auto-accept running. Press ESC to exit." << std::endl;
    std::cout << "Waiting for match..." << std::endl;

    _cancelToken.store(false);
    std::thread backgroundThread([this]() {
        this->RunAutoAcceptAsync();
    });

    while (backgroundThread.joinable()) {
        if (_kbhit()) {
            int ch = _getch();
            if (ch == 27) { // ESC key
                _cancelToken.store(true);
                break;
            }
        }
        std::this_thread::sleep_for(std::chrono::milliseconds(100));
    }

    if (backgroundThread.joinable()) {
        backgroundThread.join();
    }

    return true;
}

bool LcuClient::RunAutoAcceptAsync() {
    try {
        _matchAcceptService->RunAutoAcceptLoop(_cancelToken);
        return true;
    }
    catch (...) {
        return false;
    }
}
